dissect/target/plugins/os/windows/iis.py

```diff
@@ -167,7 +167,7 @@ def parse_w3c_format_log(self, path: Path) -> Generator[TargetRecordDescriptor,
         fields = []
         extra_fields = []
         for line in path.open().readlines():
-            line = line.decode("utf-8").strip()
+            line = line.decode("utf-8", errors="surrogateescape").strip()
```

Comment:
You're right flow.record crashes on surrogates. Changed to errors=replace. Too bad we lose the bytes though. Would be nice if there was a way to preserve them somehow, just for inspection.

Reply:
> would it be an idea to allow for surrogates in flow.record, probably not, but I ask anyway

Why not? Let's at least explore that.

Reply:
I think we better not use surrogates after all, point is, we don't know really what those bytes are. Surrogates are a technique to encode UTF-16 sequences into UTF-8, something we probably do not even want to happen since we cannot be sure it's UTF-16 either and even if it is the UTF-16 sequence might be broken (which is the case here). Surrogates are not part of the UTF-8 standard, only part of the so-called modified UTF-8 standard (but only Java uses that).

The main goal in my opinion for Dissect is to preserve data for analysis. Per UTF-8 standard we are required to dismiss all invalid byte sequences - so that is pretty simple, just replace or error out (like Python does). However, because our goal is to preserve the data, even if it is invalid (or even just because it is invalid = might be interesting), I suggest we use the "backslashreplace" option, it will just output the raw byte values which is good for inspection. The analysts can then research the data as-is.


Relevant sources:
- https://stackoverflow.com/questions/51001150/what-are-surrogate-characters-in-utf-8
- https://docs.oracle.com/javase/6/docs/api/java/io/DataInput.html#modified-utf-8

Reply:
Ah I actually thought that `backslashreplace` was the same as `surrogateescape`. What's the difference between `replace` and `backslashreplace`?

Reply:
Yeah, I was also confused. I always forget that UTF-16 somehow claimed the word 'surrogate' to mean something different that conventional language would imply.

As far as I understand:

- surrogateescape = add specific bytes that allow UTF-16/32 chars but are "invalid UTF-8" (DC80 - DCFF)
- backslashreplace = replace with \u + byte value as numerical value

Reply:
Given a function rawbytes() that just outputs the internal respresentation...

```
>>> rawbytes(b"\xa7".decode("UTF-8",errors="surrogateescape"))
b'\xdc\xa7'
>>> rawbytes(b"\xa7".decode("UTF-8",errors="backslashreplace"))
b'\\xa7'

```

Reply:
Some more nice examples:
https://johnlekberg.com/blog/2020-04-03-codec-errors.html